File: lang/python/benchmarks/benchmark_vs_sqlite.py
"""
DBX vs SQLite - Performance Comparison Benchmark

Compares DBX Python bindings with SQLite (sqlite3 module).
"""

import time
import sys
import os
import sqlite3

# Add parent directory to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from dbx_py import Database


def benchmark_dbx(n=10000):
    """Benchmark DBX operations with transactions"""
    db = Database.open_in_memory()
    
    try:
        # INSERT with transaction
        start = time.perf_counter()
        tx = db.begin_transaction()
        for i in range(n):
            key = f"key:{i}".encode()
            value = f"value:{i}".encode()
            tx.insert("bench", key, value)
        tx.commit()
        insert_time = time.perf_counter() - start
        
        # GET (no transaction needed for reads)
        start = time.perf_counter()
        for i in range(n):
            key = f"key:{i}".encode()
            _ = db.get("bench", key)
        get_time = time.perf_counter() - start
        
        # DELETE with transaction
        start = time.perf_counter()
        tx = db.begin_transaction()
        for i in range(n):
            key = f"key:{i}".encode()
            tx.delete("bench", key)
        tx.commit()
        delete_time = time.perf_counter() - start
        
        return insert_time, get_time, delete_time
    finally:
        db.close()


def benchmark_sqlite(n=10000, in_memory=True):
    """Benchmark SQLite operations with explicit transactions"""
    db_path = ":memory:" if in_memory else "sqlite_bench.db"
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    try:
        cursor.execute("CREATE TABLE IF NOT EXISTS bench (key TEXT PRIMARY KEY, value TEXT)")
        
        # INSERT with explicit transaction
        start = time.perf_counter()
        cursor.execute("BEGIN TRANSACTION")
        for i in range(n):
            key = f"key:{i}"
            value = f"value:{i}"
            cursor.execute("INSERT INTO bench (key, value) VALUES (?, ?)", (key, value))
        cursor.execute("COMMIT")
        insert_time = time.perf_counter() - start
        
        # GET (no transaction needed for reads)
        start = time.perf_counter()
        for i in range(n):
            key = f"key:{i}"
            cursor.execute("SELECT value FROM bench WHERE key = ?", (key,))
            _ = cursor.fetchone()
        get_time = time.perf_counter() - start
        
        # DELETE with explicit transaction
        start = time.perf_counter()
        cursor.execute("BEGIN TRANSACTION")
        for i in range(n):
            key = f"key:{i}"
            cursor.execute("DELETE FROM bench WHERE key = ?", (key,))
        cursor.execute("COMMIT")
        delete_time = time.perf_counter() - start
        
        return insert_time, get_time, delete_time
    finally:
        conn.close()
        if not in_memory and os.path.exists("sqlite_bench.db"):
            os.remove("sqlite_bench.db")


def print_results(name, insert_time, get_time, delete_time, n):
    """Print benchmark results"""
    print(f"\n{name}:")
    print(f"  INSERT: {insert_time:.4f}s ({n/insert_time:,.0f} ops/sec)")
    print(f"  GET:    {get_time:.4f}s ({n/get_time:,.0f} ops/sec)")
    print(f"  DELETE: {delete_time:.4f}s ({n/delete_time:,.0f} ops/sec)")


def main():
    print("=" * 60)
    print("DBX vs SQLite - Performance Comparison")
    print("=" * 60)
    
    n = 10000
    print(f"\nRunning benchmarks with {n:,} operations...\n")
    
    # Benchmark DBX
    print("Benchmarking DBX...")
    dbx_insert, dbx_get, dbx_delete = benchmark_dbx(n)
    print_results("DBX (In-Memory)", dbx_insert, dbx_get, dbx_delete, n)
    
    # Benchmark SQLite (Memory)
    print("\nBenchmarking SQLite (In-Memory)...")
    sql_insert, sql_get, sql_delete = benchmark_sqlite(n, in_memory=True)
    print_results("SQLite (In-Memory)", sql_insert, sql_get, sql_delete, n)
    
    # Benchmark SQLite (Disk)
    print("\nBenchmarking SQLite (Disk)...")
    sql_disk_insert, sql_disk_get, sql_disk_delete = benchmark_sqlite(n, in_memory=False)
    print_results("SQLite (Disk)", sql_disk_insert, sql_disk_get, sql_disk_delete, n)
    
    # Comparison
    print("\n" + "=" * 60)
    print("Performance Comparison (DBX vs SQLite In-Memory):")
    print("=" * 60)
    print(f"INSERT: DBX is {sql_insert/dbx_insert:.2f}x faster")
    print(f"GET:    DBX is {sql_get/dbx_get:.2f}x faster")
    print(f"DELETE: DBX is {sql_delete/dbx_delete:.2f}x faster")
    
    print("\n" + "=" * 60)
    print("Benchmark completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: lang/cpp/benchmarks/benchmark_vs_sqlite.cpp
/**
 * DBX vs SQLite Performance Benchmark (C++)
 * 
 * Compares DBX with SQLite using transactions for bulk operations.
 */

#include <iostream>
#include <chrono>
#include <vector>
#include <string>
#include <cstring>
#include <sqlite3.h>
#include "../include/dbx.h"

using namespace std;
using namespace chrono;

const int N = 10000;

// Helper function to format numbers with commas
string formatNumber(int num) {
    string s = to_string(num);
    int n = s.length() - 3;
    while (n > 0) {
        s.insert(n, ",");
        n -= 3;
    }
    return s;
}

// Benchmark DBX
void benchmarkDBX() {
    cout << "Benchmarking DBX (with FFI transaction)...\n\n";
    
    // Open in-memory database
    DbxHandle* db = dbx_open_in_memory();
    if (!db) {
        cerr << "Failed to open DBX database\n";
        return;
    }
    
    // INSERT with transaction
    auto startInsert = high_resolution_clock::now();
    
    DbxTransaction* tx = dbx_begin_transaction(db);
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        string value = "value:" + to_string(i);
        dbx_transaction_insert(tx, "bench", 
            (const uint8_t*)key.c_str(), key.size(),
            (const uint8_t*)value.c_str(), value.size());
    }
    dbx_transaction_commit(tx);
    
    auto endInsert = high_resolution_clock::now();
    double insertTime = duration<double>(endInsert - startInsert).count();
    
    // GET
    auto startGet = high_resolution_clock::now();
    
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        uint8_t* value = nullptr;
        size_t valueLen = 0;
        dbx_get(db, "bench", (const uint8_t*)key.c_str(), key.size(), &value, &valueLen);
        if (value) {
            dbx_free_value(value, valueLen);
        }
    }
    
    auto endGet = high_resolution_clock::now();
    double getTime = duration<double>(endGet - startGet).count();
    
    // DELETE with transaction
    auto startDelete = high_resolution_clock::now();
    
    DbxTransaction* tx2 = dbx_begin_transaction(db);
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        dbx_transaction_delete(tx2, "bench", (const uint8_t*)key.c_str(), key.size());
    }
    dbx_transaction_commit(tx2);
    
    auto endDelete = high_resolution_clock::now();
    double deleteTime = duration<double>(endDelete - startDelete).count();
    
    dbx_close(db);
    
    // Print results
    cout << "DBX (In-Memory, FFI Transaction):\n";
    cout << "  INSERT: " << insertTime << "s (" << formatNumber((int)(N / insertTime)) << " ops/sec)\n";
    cout << "  GET:    " << getTime << "s (" << formatNumber((int)(N / getTime)) << " ops/sec)\n";
    cout << "  DELETE: " << deleteTime << "s (" << formatNumber((int)(N / deleteTime)) << " ops/sec)\n\n";
}

// Benchmark SQLite
void benchmarkSQLite() {
    cout << "Benchmarking SQLite (In-Memory)...\n\n";
    
    sqlite3* db;
    if (sqlite3_open(":memory:", &db) != SQLITE_OK) {
        cerr << "Failed to open SQLite database\n";
        return;
    }
    
    // Create table
    sqlite3_exec(db, "CREATE TABLE bench (key TEXT PRIMARY KEY, value TEXT)", nullptr, nullptr, nullptr);
    
    // INSERT with transaction
    auto startInsert = high_resolution_clock::now();
    
    sqlite3_exec(db, "BEGIN TRANSACTION", nullptr, nullptr, nullptr);
    
    sqlite3_stmt* insertStmt;
    sqlite3_prepare_v2(db, "INSERT INTO bench (key, value) VALUES (?, ?)", -1, &insertStmt, nullptr);
    
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        string value = "value:" + to_string(i);
        sqlite3_bind_text(insertStmt, 1, key.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(insertStmt, 2, value.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_step(insertStmt);
        sqlite3_reset(insertStmt);
    }
    
    sqlite3_finalize(insertStmt);
    sqlite3_exec(db, "COMMIT", nullptr, nullptr, nullptr);
    
    auto endInsert = high_resolution_clock::now();
    double insertTime = duration<double>(endInsert - startInsert).count();
    
    // GET
    auto startGet = high_resolution_clock::now();
    
    sqlite3_stmt* getStmt;
    sqlite3_prepare_v2(db, "SELECT value FROM bench WHERE key = ?", -1, &getStmt, nullptr);
    
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        sqlite3_bind_text(getStmt, 1, key.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_step(getStmt);
        sqlite3_reset(getStmt);
    }
    
    sqlite3_finalize(getStmt);
    
    auto endGet = high_resolution_clock::now();
    double getTime = duration<double>(endGet - startGet).count();
    
    // DELETE with transaction
    auto startDelete = high_resolution_clock::now();
    
    sqlite3_exec(db, "BEGIN TRANSACTION", nullptr, nullptr, nullptr);
    
    sqlite3_stmt* deleteStmt;
    sqlite3_prepare_v2(db, "DELETE FROM bench WHERE key = ?", -1, &deleteStmt, nullptr);
    
    for (int i = 0; i < N; i++) {
        string key = "key:" + to_string(i);
        sqlite3_bind_text(deleteStmt, 1, key.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_step(deleteStmt);
        sqlite3_reset(deleteStmt);
    }
    
    sqlite3_finalize(deleteStmt);
    sqlite3_exec(db, "COMMIT", nullptr, nullptr, nullptr);
    
    auto endDelete = high_resolution_clock::now();
    double deleteTime = duration<double>(endDelete - startDelete).count();
    
    sqlite3_close(db);
    
    // Print results
    cout << "SQLite (In-Memory):\n";
    cout << "  INSERT: " << insertTime << "s (" << formatNumber((int)(N / insertTime)) << " ops/sec)\n";
    cout << "  GET:    " << getTime << "s (" << formatNumber((int)(N / getTime)) << " ops/sec)\n";
    cout << "  DELETE: " << deleteTime << "s (" << formatNumber((int)(N / deleteTime)) << " ops/sec)\n\n";
}

int main() {
    cout << "============================================================\n";
    cout << "DBX vs SQLite - Performance Comparison (C++)\n";
    cout << "============================================================\n\n";
    cout << "Running benchmarks with " << formatNumber(N) << " operations...\n\n";
    
    benchmarkDBX();
    benchmarkSQLite();
    
    cout << "============================================================\n";
    cout << "Benchmark completed!\n";
    cout << "============================================================\n";
    
    return 0;
}
